trulens/utils/typing.py

```diff
@@ -5,19 +5,212 @@
 
 from dataclasses import dataclass, field
 from inspect import signature
-from typing import Any, Callable, Dict, List, Optional, Union
+from typing import (
+    Any, Callable, Dict, Generic, Iterable, List, Optional, Tuple, TypeVar,
+    Union)
 
 import numpy as np
+from trulens.nn.backend import Tensor
 
-ArrayLike = Union[np.ndarray, Any, List[Union[np.ndarray, Any]]]
+# Atomic model inputs (at least from our perspective)
+DataLike = Union[np.ndarray, Tensor]
 
+# Model input arguments. Either a single element or list/tuple of several.
+ArgsLike = Union[DataLike, List[DataLike], Tuple[DataLike]]
+
+# Model input kwargs.
+Kwargs = Dict[str, DataLike]
+
+# Some backends allow inputs in terms of dicts with tensors as keys (i.e. feed_dict in tf1).
+# We keep track of internal names so that these feed dicts can use strings as names instead.
+Feed = Dict[Union[str, Tensor], DataLike]
+
+KwargsLike = Union[Kwargs, Feed]
+
+C = TypeVar("C")
+C1 = TypeVar("C1")
+C2 = TypeVar("C2")
+K = TypeVar("K")
+V = TypeVar("V")
+
+Indexable = Union[List[V], Tuple[V]]
+# For type checking the above.
 DATA_CONTAINER_TYPE = (list, tuple)
 
 
+# Convert a single element input to the multi-input one.
+def as_args(ele):
+    if isinstance(ele, DATA_CONTAINER_TYPE):
+        return ele
+    else:
+        return [ele]
+
+
+class IndexableUtils:
+
+    def with_(l: Indexable[V], i: int, v: V) -> Indexable[V]:
+        """Copy of the given list or tuple with the given index replaced by the given value."""
+
+        if isinstance(l, list):
+            l = l.copy()
+            l[i] = v
+        elif isinstance(l, tuple):
+            l = list(l)
+            l[i] = v
+            l = tuple(l)
+        else:
+            raise ValueError(
+                f"list or tuple expected but got {l.__class__.__name__}")
+
+        return l
+
+
+class IterableUtils:
+
+    def then_(iter1: Iterable[V], iter2: Iterable[V]) -> Iterable[V]:
+        """Iterate through the given iterators, one after the other."""
+
+        for x in iter1:
+            yield x
+        for x in iter2:
+            yield x
+
+
+class DictUtils:
+
+    def with_(d: Dict[K, V], k: K, v: V) -> Dict[K, V]:
+        """Copy of the given dictionary with the given key replaced by the given value."""
+
+        d = d.copy()
+        d[k] = v
+        return d
+
+
+@dataclass
+class Lens(Generic[C, V]):
+    """Simple lenses implementation."""
+
+    get: Callable[[C], V]
+    # Given a container, extract the focused value.
+
+    set: Callable[[C, V], C]
+    # Given a container and a value, replace the focused value with the new one,
+    # returning a new container.
+
+    @staticmethod
+    def lenses_elements(c: List[V]) -> Iterable['Lens[List[V], V]']:
+        """Lenses focusing on elements of a list."""
+
+        for i in range(len(c)):
+            yield Lens(
+                lambda l, i=i: l[i],
+                lambda l, v, i=i: IndexableUtils.with_(l, i, v))
```

Comment:
why do we need this class?

Reply:
For unifying access to different type of args in ModelInputs .

Reply:
resolved, just add a link to Lenses
maybe clarify C-> container, V->value
and that this is to help treat args/kwargs similarly as a convenience

Reply:
Done.